Scenario: We are given a rope of 1 unit length. We decide to cut this rope into three parts. What is the probability that the three cut parts successfully form three sides of a triangle? 

For a triangle the following property holds: The sum of any two sides of a triangle is always greater than its third side. 

Note that the question itself is not difficult, but I would like to motivate here the significance of simulation based fact learning.

### Function 1

We define a function namely check () that takes as input the dimensions of three cut parts and state if they can successfully form sides of a triangle.

In [1]:
def check (first, second, third):
    if ((first + second > third) and (second + third > first) and (first + third > second)): # If these three cut rope parts 
        # successfully form sides of a triangle
        return True
    # The function defaults to returning false
    return False

Let's invoke the check function on some inputs and see how it behaves.

In [2]:
check (3,4,5)

True

In [3]:
check (2, 7, 8)

True

In [4]:
check (1, 1, 1) # Equilateral triangle

True

In [5]:
check (4, 1, 3)

False

In [6]:
check (4, 4, 1) # Isosceles triangle

True

Now, a question arises, how should I break the three ropes from a unit length rope? In order to avoid complications that arise from decimal numbers, let us instead scale the rope length to 10.

In [7]:
rope_length = 10

We must then ensure that all three cut parts add up to rope_length. 

We will use the naive mathematical understanding of probability that defines it as favorable outcomes (successfully forming sides of a triangle) to total outcomes (total possible ways of legitimately cutting the rope into three parts)

### Function 2

We define a function namely probability () that takes as input rope length and returns the probability that three cut parts from the rope successfuly form three sides of a triangle. Note that this formulation of the function probability () that takes rope length as input is essential for the final point that I wish to make regarding simulation based fact learning.

In [8]:
def probability (rope_length): 
    total_outcomes = 0 # A legitimate three piece cut forms part of the total outcomes
    favorable_outcomes = 0 # Three cut parts successfully forming three sides of a triangle form part of favorable outcome
    for first in range (1, rope_length): # This range can be smaller as well but I am keeping things simple
        for second in range (1, rope_length):
            for third in range (1, rope_length):
                if (first + second + third) == rope_length:
                    total_outcomes += 1
                    if (check (first, second, third)): # Only if it is a legitimate cut, check if it is also a successful cut
                        favorable_outcomes += 1
    return (favorable_outcomes/ total_outcomes)

In [9]:
probability (rope_length) * 100

16.666666666666664

Note that we get a probability of 16.67%. However, this isn't the correct probability. The scaling of rope length to 10 is not sufficient to reach the correct fact. So, one strategy that we can follow is increase the scaling of rope_length in and see if the result converges to some stable probability value.

In [10]:
rope_length = 50
probability (rope_length) * 100

23.46938775510204

In [11]:
rope_length = 100
probability (rope_length) * 100

24.242424242424242

In [12]:
rope_length = 500
probability (rope_length) * 100

24.849699398797593

In [13]:
rope_length = 1000
probability (rope_length) * 100

24.924924924924923

In [14]:
rope_length = 1500 # This takes a long time to run
probability (rope_length) * 100

24.94996664442962

Even though it is not an optimized program and has a very high running time complexity; it however, does illustrate that the answer converges to approximately 25% which is the correct answer.